projects/ledger-app/0001-token-admin-methods.md

```diff
@@ -0,0 +1,219 @@
+[guide-level-explanation]: #guide-level-explanation
+
+> [!WARNING] Request for comment
+> This whole paragraph below can change if we really want to, we just need to add an `isNFT` boolean in the token data for the user to confirm, so in all transactions the Ledger will be informed that the token is an NFT with the user express consent during token signing
+> This makes a breaking change in the wallet dekstop since we will change an existing command.
+> We can avoid this command change by using the P1-P2 to create a SEND_TOKEN_DATA v2 where we have the `isNFT` and keep the v1 unchamged.
+
+Before entering the operations, we should address the differences between custom tokens and NFTs.
+In Hathor, NFTs are custom tokens, where the creation transaction (which is the uid of the NFT) has the NFT data associated which is usually an IPFS link to the NFT document (image, pdf, video, etc.).
+After the creation the operations and transactions involving NFT are exactly the same regarding the protocol, but the balance is shown differently to the user, where the custom token unit is worth 0,01 the NFT is always integer, so it is 1. Both of these are the same value on the transaction it is just shown different on the UI.
+Since the Ledger cannot check if a token is an NFT or not, we should disregard any changes and treat both as custom tokens (which is valid if we follow the protocol).
+
+## Transaction parsing
+
+To allow admin operations we will introduce new concepts into the app that were being ignored or not treated.
+
+---
+
+The `token_data` of an output is a byte with the information:
+- Weather the token is a custom token or HTR
+- Which custom token it is (index on the token array) 
+- If the output is an authority output (which authority will be encoded on the `value`)
+Currently on the app we ignore the last piece of information since we do not allow authority outputs.
+The way we do it is by throwing an error of the `token_data & TOKEN_DATA_AUTHORITY_MASK` is not 0.
+To allow authority outputs we just need to change this and create a special screen for authority outputs when the user is confirming them.
+Usually the output confirmation screen involves 4 steps:
+- Address step (with pagination)
+- Amount step (Showing as `ABC 0.01` or `HTR 0.01`, with pagination for big values)
+- Approve step
+- Reject step
+
+When confirming an authority output we should show:
+- Address step (with pagination)
+- Type step (`mint authority` or `melt authority`)
+	- Other authorities will throw.
+	- We will not allow joined authority, for instance `mint+melt authority` (i.e. `value = 3`)
+- Approve step
+- Reject step
+
+Obs: With authority outputs we already have enough to enable mint, melt, delegate and destroy operations.
+
+---
+
+Currently we only allow P2PKH scripts (and addresses) but NFTs have a data output.
+Data outputs are allowed on the protocol but we will only allow on NFT creation scripts because any other is just an HTR burn.

```

Comment:
Do we want to allow any data output on any transaction?
If so I can change it.

I just think that the app should be restrictive on what we want to allow, that's why I added this but as stated in the design there is no issue either on the app or the protocol to add support for data outputs.

If we want to add support for data outputs we just need to confirm which encoding we want to show the user.
The NFT data output would use ascii (subset of UTF-8) so the user con confirm the NFT url but if its not an NFT creation I think we should use hex encoding (to allow any byte string).
What do you think?

Obs: data scripts are defined as scripts that could not be parsed, meaning they are not P2SH or P2PKH. We currently do not check for P2SH so to add support for all data outputs we would need to add P2SH parsing, unless we accept a P2SH script showing as a data script until we add support for P2SH.

Reply:
> We have a way to parse the output script to identify its type, we could use it, can't we?

But we dont have a way to identify the script, we would need to implement one.
The idea would be to check if its P2PKH, then check P2SH and if none of the above match, it is a data output.
The "check if its P2SH" method does not exist currently.

> my main concern here is not adding this support but that the code to handle the NFT data output is not too specific

I think making it very specific is what gives us more flexibility in the future.
The case of the design is only for "Create token transactions" and only for the first data output on them.
In the future we can add support for data outputs in other places since it won't interfere with this specific case.